### PR TITLE
feat: hint

### PR DESCRIPTION
**Pull request by JimmyDaddy**
* [optimizer-hints](https://dev.mysql.com/doc/refman/5.7/en/optimizer-hints.html)
* [index-hints](https://dev.mysql.com/doc/refman/5.7/en/index-hints.html)
* [mssql-hints](https://docs.microsoft.com/en-us/sql/t-sql/queries/hints-transact-sql?view=sql-server-ver15) not support

### MySQL optimize hint
```js
// pass hint object
Post.create({ title: 'New Post', createdAt: date, updatedAt: date }, { hint: new Hint('SET_VAR(foreign_key_checks=OFF)') })
// INSERT /*+ SET_VAR(foreign_key_checks=OFF) */ INTO `articles` (`gmt_create`, `gmt_modified`, `title`) VALUES ('2017-12-12 00:00:00.000
// pass plain object
Post.create({ title: 'New Post', createdAt: date, updatedAt: date }, { hint: { value: 'SET_VAR(foreign_key_checks=OFF) }} })
// INSERT /*+ SET_VAR(foreign_key_checks=OFF) */ INTO `articles` (`gmt_create`, `gmt_modified`, `title`) VALUES ('2017-12-12 00:00:00.000
```
### MySQL index hint
```js

Post.update({
    title: { $like: '%Post%' }
  },
  { title: 'hello', updatedAt: date },
  { 
      hints: [ 
          new IndexHint('idx_id'),
         { value: 'idx_title', type: INDEX_HINT_TYPE.USE }, 
         new IndexHint('idx_title') 
     ] 
});
// UPDATE `articles` USE INDEX (idx_id,idx_title) SET `title` = 'hello', `gmt_modified` = '2017-12-12 00:00:00.000' WHERE `title` LIKE '%Post%' AND `gmt_deleted` IS NULL

// IGNORE|FORCE|USE INDEX FOR (**)
Post.find({ title: { $like: '%Post%' } }, { order: 'id', hints: [ new IndexHint('idx_id', INDEX_HINT_TYPE.USE, INDEX_HINT_USE_TYPE.ORDER_BY), new IndexHint('idx_title'), new IndexHint('idx_title') ] })
// SELECT * FROM `articles` USE INDEX FOR ORDER BY (idx_id) USE INDEX (idx_title) WHERE `title` LIKE '%Post%' AND `gmt_deleted` IS NULL ORDER BY `id`

```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -11,6 +11,7 @@ const migrations = require('./lib/migrations');
 const sequelize = require('./lib/adapters/sequelize');
 const { camelCase } = require('./lib/utils/string');
 const Spell = require('./lib/spell');
+const Hint = require('./lib/hint');
```
Comment (cyjake) [inline index.js:14]: aha, totally forgot this feature, nice pick.

File: lib/hint.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+// MSSQL only, not support yet
+class TableHint extends Hint {
+  constructor(value) {
+    super(value, HINT_TYPE.TABLE_HINT)
+  }
+}
+
+// MYSQL only
```
Comment (cyjake) [inline lib/hint.js:47]: MySQL

File: lib/hint.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    return '';
+  }
+}
+
+// MSSQL only, not support yet
+class TableHint extends Hint {
+  constructor(value) {
+    super(value, HINT_TYPE.TABLE_HINT)
+  }
+}
```
Comment (cyjake) [inline lib/hint.js:45]: 👍

File: lib/hint.js [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+  // static
+  /**
+   *
+   *
+   * @static
+   * @param {IndexHint} hints
+   * @returns {Array<IndexHint>}
+   * @memberof IndexHint
+   */
+  static merge(hints) {
```
Comment (JimmyDaddy) [inline lib/hint.js:179]: merge index hint by type and values

File: lib/hint.js [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+      }
+      groupArr.push(newIndexHint);
+    }
+
+    return groupArr;
+  }
+}
+
+// simulate interface
+function buildHint(obj) {
```
Comment (JimmyDaddy) [inline lib/hint.js:209]: build hint with plain object

File: lib/drivers/mysql/spellbook.js [hunk 6]
```diff
@@ -48,9 +50,18 @@ function formatMysqlUpsert(spell) {
...
+    if (!spell || !spell.optimizeHint.length) return '';
+    const hintStr = spell.optimizeHint.map(hint => hint.toSqlString()).join(' ') || '';
+    return `/*+ ${hintStr} */`;
+  },
+
+  formatIndexHint(spell) {
+    if (!spell || !spell.indexHint.length) return '';
+    // merge index hint
+    const hints = IndexHint.merge(spell.indexHint);
+    const hintStr = hints.map(hint => hint.toSqlString()).join(' ') || '';
```
Comment (cyjake) [inline lib/drivers/mysql/spellbook.js:63]: `[].join(' ')` returns `''`, the extra `|| ''` looks unnecessary

File: lib/spell.js [hunk 7]
```diff
@@ -481,6 +482,7 @@ class Spell {
       throw new Error(`model ${Model.name} is not connected yet`);
     }
 
+
```
Comment (cyjake) [inline lib/spell.js:485]: redundant empty line

File: lib/spell.js [hunk 8]
```diff
@@ -493,15 +495,17 @@ class Spell {
       scopes.push(scopeDeletedAt);
     }
 
-    this.hints = opts.hints || [];
+    if (opts.scopes && opts.scopes.length) {
+      scopes.concat(opts.scopes);
```
Comment (cyjake) [inline lib/spell.js:499]: `scopes = scopes.concat(opts.scopes)` or `scropts.push(...opts.scopes)`

File: test/unit/hint.test.js [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+            new IndexHint('idx_title', INDEX_HINT_TYPE.FORCE),
+            new IndexHint('idx_title', INDEX_HINT_TYPE.FORCE),
+            new IndexHint('idx_hle', INDEX_HINT_TYPE.IGNORE),
+            new IndexHint('idx_hle1', INDEX_HINT_TYPE.IGNORE),
+            new IndexHint('idx_hle', INDEX_HINT_TYPE.IGNORE),
+            new IndexHint('idx_hle2', INDEX_HINT_TYPE.USE),
+            new Hint('SET_VAR(foreign_key_checks=OFF)'),
+            new Hint('MAX_EXECUTION_TIME(1000)')
+          ]
+        }).toString(),
```
Comment (cyjake) [inline test/unit/hint.test.js:199]: The indentation of this block looks a bit weird. Suggestion:

```js
Post.update(
  { title: { $like: '%Post%' } },
  { title: 'hello', updatedAt: date },
  {
    hints: [
      new IndexHint('idx_id', INDEX_HINT_TYPE.FORCE),
      new IndexHint('idx_title', INDEX_HINT_TYPE.FORCE),
      new IndexHint('idx_title', INDEX_HINT_TYPE.FORCE),
      new IndexHint('idx_hle', INDEX_HINT_TYPE.IGNORE),
      new IndexHint('idx_hle1', INDEX_HINT_TYPE.IGNORE),
      new IndexHint('idx_hle', INDEX_HINT_TYPE.IGNORE),
      new IndexHint('idx_hle2', INDEX_HINT_TYPE.USE),
      new Hint('SET_VAR(foreign_key_checks=OFF)'),
      new Hint('MAX_EXECUTION_TIME(1000)')
    ]
  }
).toString(),
```

Comment (cyjake) [reply test/unit/hint.test.js:199]: Personally I'd prefer method chaining style like Active Record https://bigbinary.com/blog/rails-6-supports-optimizer-hints

```js
Post.update({ title: { $like: '%Post%' } }, { title: 'hello', updatedAt: date })
  .optimizerHints('SET_VAR(foreign_key_checks=OFF)', 'MAX_EXECUTION_TIME(1000)')
  .useIndex('idx_hle2')
  .forceIndex('idx_id', 'idx_title', 'idx_title')
  .ignoreIndex('idx_hle', 'idx_hle1', 'idx_hle')
```

File: lib/hint.js (hunk 5 above)
Comment (cyjake) [reply lib/hint.js:209]: a better place to put this method is to define it as a static method of `Hint`

```js
class Hint {
  static build() {}
}
```

File: test/unit/hint.test.js [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+      assert.equal(
+        Post.find({
+          title: { $like: '%Post%' } },
+          {
+            order: 'id',
+            hints: [
+              new IndexHint('idx_id', INDEX_HINT_TYPE.FORCE),
+              new IndexHint('idx_title', INDEX_HINT_TYPE.FORCE),
+              new IndexHint('idx_title', INDEX_HINT_TYPE.FORCE),
+              new IndexHint('idx_title', INDEX_HINT_TYPE.FORCE, INDEX_HINT_USE_TYPE.ORDER_BY),
```
Comment (cyjake) [inline test/unit/hint.test.js]:
```js
Post.order('id').forceIndex({ orderBy: 'idx_title' });
```

File: test/unit/realm.test.js [hunk 11]
```diff
@@ -127,14 +128,15 @@ describe('=> Realm', () => {
...
-        const sql = 'INSERT INTO users (gmt_create, email, nickname) VALUES (?, ?, ?)';
-        await realm.query(sql, [ new Date(), email, 'Thor' ], { connection });
-        await realm.query(sql, [ new Date(), email, 'Loki' ], { connection });
-      }, `Error: ER_DUP_ENTRY: Duplicate entry '${email}' for key 'users.email'`); // rollback
-    });
+        const sql = 'INSERT INTO users (gmt_create, email, nickname, status) VALUES (?, ?, ?, ?)';
+        await realm.query(sql, [ new Date(), email, 'Thor', 1 ], { connection });
+        await realm.query(sql, [ new Date(), email, 'Loki', 1 ], { connection });
+      }); // rollback
+    }, `Error: ER_DUP_ENTRY: Duplicate entry '${email}' for key 'users.email'`);
```
Comment (cyjake) [inline test/unit/realm.test.js:137]: 😅

File: test/unit/hint.test.js [hunk 12]
```diff
@@ -100,21 +124,55 @@ describe('MySQL', async () => {
...
+            new IndexHint('idx_title', INDEX_HINT_TYPE.FORCE, INDEX_HINT_USE_TYPE.ORDER_BY), // USE INDEX FOR ** ()
+            new IndexHint('idx_hle', INDEX_HINT_TYPE.IGNORE),
+            new IndexHint('idx_hle1', INDEX_HINT_TYPE.IGNORE),
+            new IndexHint('idx_hle', INDEX_HINT_TYPE.IGNORE),
+            new IndexHint('idx_hle2', INDEX_HINT_TYPE.USE),
+          ]
+        })
+        .forceIndex('idx_id', 'idx_title')
+        .forceIndex('idx_title')
+        .forceIndex({ value: 'idx_title', useFor: 'ORDER BY' })
```
Comment (cyjake) [inline test/unit/hint.test.js:171]: still prefer `.forceIndex({ orderBy: 'idx_title' })`

**Pull request merged**